protocol/contracts/nexus/Shard.sol

```diff
@@ -0,0 +1,555 @@
+
+    /// @notice minters who can mint for templars
+    mapping(address => bool) public templarMinters;
+
+    mapping(uint256 => Recipe) private recipes;
+
+    /// @notice each shard belongs to exactly 1 enclave. an enclave can have many shards
+    mapping(uint8 => EnumerableSet.UintSet) private enclaveShards;
+
+    /// @notice Enclave types
+    enum Enclave {
+        Chaos,
+        Mystery,
+        Logic,
+        Order,
+        Structure
+    }
+
+    /// @notice for compact view functions
+    struct PartnerMintInfo {
+        uint256[] shardIds;
+        uint256[] caps;
+        uint256[] balances;
+    }
+
+    /// @dev maintaining uint256[] to avoid casting in batch functions like burnBatch
+    struct Recipe {
+        uint256[] inputShardIds;
+        uint256[] inputShardAmounts;
+        uint256[] outputShardIds;
+        uint256[] outputShardAmounts;
+    }
+
+    event Transmuted(address caller, uint256 recipeId);
+    event ShardMinted();
+    event ShardUriSet(uint256 shardId, string uri);
+    event RecipeSet(uint256 recipeId, Recipe recipe);
+    event PartnerAllowedShardIdSet(address partner, uint256 shardId, bool allow);
+    event PartnerAllowedShardIdsSet(address partner, uint256[] shardIds, bool[] allowances);
+    event PartnerAllowedShardCapsSet(address partner, uint256[] shardIds, uint256[] caps);
+    event TemplarMinterSet(address minter, bool allowed);
+    event RecipeDeleted(uint256 recipeId);
+    event ShardEnclaveSet(Enclave enclave, uint256 shardId);
+    event RegularShardIdSet(uint256 shardId, bool allow);
+
+    error ReservedPartnerShard(uint256 shardId);
+    error ShardMintNotAllowed(address caller, uint256 shardId);
+    error CannotTransmute(address caller);
+    error CannotMint(address caller);
+    error MintCapExceeded(uint256 cap, uint256 amount);
+    error InvalidRecipe(uint256 recipeId);
+    error ZeroAddress();
+    error InvalidParamLength();
+    error InvalidAccess(address caller);
+    error InvalidCaller(address caller);
+    error ERC1155MissingApprovalForAll(address msgSender, address account);
+    error AccountBlacklisted(address account);
+    error SetShardEnclaveFailed();
+    error InvalidShard(uint256 shardId);
+
+
+    constructor(
+        address _relic,
+        address _initialRescuer,
+        address _initialExecutor,
+        string memory _uri
+    ) ERC1155(_uri) TempleElevatedAccess(_initialRescuer, _initialExecutor) {
+        relic = IRelic(_relic);
+    }
+
+    /*
+     * Set shard enclave
+     * @param enclave Enclave
+     * @param shardId Shard ID
+     */
+    function setShardEnclave(Enclave enclave, uint256 shardId) external onlyElevatedAccess {
+        /// remove if shard already belongs to an enclave
+        uint256 _length = uint256(uint8(Enclave.Structure)) + 1;
+        uint8 enclaveIndex = uint8(enclave);
+        for (uint i; i < _length;) {
+            if (enclaveShards[uint8(i)].contains(shardId)) {
```

Comment:
No need for this `if` check, since `remove()` will internally check if it's there or not

Reply:
there's a `break` to break out of the loop if condition is true. without this, worst case, loop will call `remove()` `_length` times.